src/java.base/share/classes/sun/security/provider/HSS.java

```diff
@@ -0,0 +1,753 @@
+        messageStream.write(data, off, len);
+    }
+
+    protected boolean engineVerify(byte[] signature) throws SignatureException {
+        HSSSignature sig = new HSSSignature(signature, pubKey);
+        LMSPublicKey lmsPubKey = pubKey.lmsPublicKey;
+        boolean result = true;
+        try {
+            for (int i = 0; i < sig.Nspk; i++) {
+                byte[] keyArr = sig.pubList[i].keyArray();
+                result &= LMSUtils.lmsVerify(lmsPubKey, sig.siglist[i], keyArr);
+                lmsPubKey = sig.pubList[i];
+            }
+
+            result &= LMSUtils.lmsVerify(lmsPubKey, sig.siglist[sig.Nspk], messageStream.toByteArray());
+        } finally {
+            messageStream.reset();
+        }
+        return result;
+    }
+
+    static class LMSPublicKey {
+        final int type;
+        final int otsType;
+        final LMSParams lmsParams;
+        private final byte[] I;
+        private final byte[] T1;
+
+        LMSPublicKey(byte[] keyArray, int offset, boolean checkExactLength) throws InvalidKeyException {
+            int inLen = keyArray.length - offset;
+            if (inLen < 8) {
+                throw new InvalidKeyException("LMS public key is too short");
+            }
+            type = LMSUtils.fourBytesToInt(keyArray, offset);
+            otsType = LMSUtils.fourBytesToInt(keyArray, offset + 4);
+            LMOTSParams lmotsParams;
+
+            try {
+                lmsParams = LMSParams.of(type);
+                lmotsParams = LMOTSParams.of(otsType);
+            } catch (IllegalArgumentException e) {
+                throw new InvalidKeyException(e.getMessage());
+            }
+
+            int m = lmsParams.m;
+            if ((inLen < (24 + m)) || (checkExactLength && (inLen != (24 + m))) ||
+                    !lmsParams.hasSameHash(lmotsParams)) {
+                throw new InvalidKeyException("Wrong LMS public Key length");
+            }
+
+            I = Arrays.copyOfRange(keyArray, offset + 8, offset + 8 + 16);
+            T1 = Arrays.copyOfRange(keyArray, offset + 24, offset + 24 + m);
+        }
+
+        void getI(byte[] arr, int pos) {
+            System.arraycopy(I, 0, arr, pos, 16);
+        }
+
+        boolean isT1(byte[] arr, int pos) {
+            int m = lmsParams.m;
+            int diff = 0;
+            for (int i = 0; i < m; i++) {
+                diff |= (T1[i] ^ arr[pos + i]);
+            }
+            return (diff == 0);
+        }
+
+        byte[] keyArray() {
+            byte[] result = new byte[keyArrayLength()];
+            LMSUtils.intToFourBytes(type, result, 0);
+            LMSUtils.intToFourBytes(otsType, result, 4);
+            System.arraycopy(I, 0, result, 8, 16);
+            System.arraycopy(T1, 0, result, 24, lmsParams.m);
+            return result;
+        }
+
+        int keyArrayLength() {
+            return 24 + lmsParams.m;
+        }
+    }
+
+    static class LMSUtils {
```

Comment:
It might be simpler and more efficient to use enum for lms and lmots types.

Reply:
I had considered that and decided not to use it. In my opinion, Java Enum is much more complicated than it should be for this case. Efficiency is not a concern here, but I also don't see how enum could be more efficient.

Reply:
> I had considered that and decided not to use it. In my opinion, Java Enum is much more complicated than it should be for this case. 

OK.

> Efficiency is not a concern here

OK.

> but I also don't see how enum could be more efficient.

enum switch is O(1), while if-else not.